q2cli/handlers.py

```diff
@@ -553,6 +558,8 @@ def get_value(self, arguments, fallback=None):
             if type(value) is not bool:
                 value = self._parse_boolean(value)
             return value
+        elif self.ast['name'] == 'Set':
+            return set(value)
```

Comment:
`Set` is a generic and has a parameterized type: `Set[Str % Choices({'foo', 'bar'})]`, `Set[Int]`, etc. So we'd probably need to dispatch the handler again for every element in `value` on the contained type (`Str % Choices({'foo', 'bar'})`, `Int`).

Reply:
Testing it with the `q2-diversity` PR that uses `Set[Str % Choices(...)]` worked as expected. I hadn't gotten a chance to write any actual tests for it though.

Reply:
```bash
(qiime2) caporasolab $ qiime diversity alpha-rarefaction --i-table
~/Desktop/qiime2-moving-pictures-tutorial/table.qza --i-phylogeny
~/Desktop/qiime2-moving-pictures-tutorial/rooted-tree.qza --p-max-depth 10000
--o-visualization test --p-metric faith_pd --p-metric pielou_e
Saved Visualization to: test.qzv
(qiime2) caporasolab $ qiime diversity alpha-rarefaction --i-table
~/Desktop/qiime2-moving-pictures-tutorial/table.qza --i-phylogeny
~/Desktop/qiime2-moving-pictures-tutorial/rooted-tree.qza --p-max-depth 10000
--o-visualization test --p-metric faith_pd --p-metric pielou_e --p-metric error
Plugin error from diversity:

  Argument to parameter 'metric' is not a subtype of Set[Str %
  Choices({'ace', 'berger_parker_d', 'brillouin_d', 'chao1',
  'dominance', 'doubles', 'enspie', 'faith_pd', 'fisher_alpha',
  'gini_index', 'goods_coverage', 'heip_e', 'lladser_pe', 'margalef',
  'mcintosh_d', 'mcintosh_e', 'menhinick', 'michaelis_menten_fit',
  'observed_otus', 'pielou_e', 'robbins', 'shannon', 'simpson',
  'simpson_e', 'singles'})].

Debug info has been saved to /var/folders/zc/30qh4_v57slc7nh0g9xqw7lh0000gn/T/qiime2-q2cli-err-nxeqen55.log
(qiime2) caporasolab $ 
```

Reply:
:+1: 

I bet we can get Click to give a clearer error message if the option is declared as a "multi-option" (`multiple=True`) of `click.Choices`. Would you mind testing that out?

Reply:
Yeah I noticed it doesn't clarify. I can dig into it!

Reply:
I suspect you will need to kind of refactor the way `RegularParameterHandler` works (this current implementation has always been a hack). You'll need some kind of way to build up the click option that should be returned by walking through the primitive type. (And then do the same thing when getting a value, so that the following test works:

User passes
```bash
--p-some-set-input 1 --p-some-set-input 2 --p-some-set-input 3
```
Where  the type is: 
```
some_set_input: Set[Int]
```
So that q2cli returns `{1, 2, 3}` and not `{"1", "2", "3"}` which is what I _think_ the current code does, since it and click are unaware of the `Int` inside of `Set`.

Reply:
Oh ok. I see what you mean now. QIIME 2 was correctly handling choices, etc but click will cast to string no matter what at the moment.